src/Components/Server/src/Circuits/CircuitRegistry.cs

```diff
@@ -0,0 +1,243 @@
+    /// <see cref="DisconnectedCircuits"/> where it sits with an expiration time. We'll mark the associated <see cref="CircuitClientProxy"/> as disconnected
+    /// so that consumers of the Circuit know of the current state. 
+    /// Once the entry for the host in <see cref="DisconnectedCircuits"/> expires, we'll dispose off the host.
+    /// 
+    /// The alternate case is when the disconnect was transient, e.g. due to a network failure, and the client attempts to reconnect.
+    /// We'll attempt to connect it back to the host and the preserved server state, when available. In this event, we do the opposite of
+    /// what we did during disconnect - transition the host from <see cref="DisconnectedCircuits"/> to <see cref="ConnectedCircuits"/>, and transfer
+    /// the <see cref="CircuitClientProxy"/> to use the new client instance that attempted to reconnect to the server. Removing the entry from 
+    /// <see cref="DisconnectedCircuits"/> should ensure we no longer have to concern ourselves with entry expiration.
+    ///
+    /// Knowing when a client disconnected is not an exact science. There's a fair possiblity that a client may reconnect before the server realizes.
+    /// Consequently, we have to account for reconnects and disconnects occuring simultaneously as well as appearing out of order.
+    /// To manage this, we use a critical section to manage all state transitions.
+    /// </remarks>
+    internal class CircuitRegistry
+    {
+        private readonly object CircuitRegistryLock = new object();
+        private readonly ComponentsServerOptions _options;
+        private readonly ILogger _logger;
+        private readonly PostEvictionCallbackRegistration _postEvictionCallback;
+
+        public CircuitRegistry(
+            IOptions<ComponentsServerOptions> options,
+            ILogger<CircuitRegistry> logger)
+        {
+            _options = options.Value;
+            _logger = logger;
+
+            ConnectedCircuits = new ConcurrentDictionary<string, CircuitHost>(StringComparer.Ordinal);
+
+            DisconnectedCircuits = new MemoryCache(new MemoryCacheOptions
+            {
+                SizeLimit = _options.MaxRetainedDisconnectedCircuits,
+            });
+
+            _postEvictionCallback = new PostEvictionCallbackRegistration
+            {
+                EvictionCallback = OnEntryEvicted,
+            };
+        }
+
+        internal ConcurrentDictionary<string, CircuitHost> ConnectedCircuits { get; }
+
+        internal MemoryCache DisconnectedCircuits { get; }
+
+        /// <summary>
+        /// Registers an active <see cref="CircuitHost"/> with the register.
+        /// </summary>
+        public void Register(CircuitHost circuitHost)
+        {
+            if (!ConnectedCircuits.TryAdd(circuitHost.CircuitId, circuitHost))
+            {
+                throw new ArgumentException($"Circuit with identity {circuitHost.CircuitId} is already registered.");
```

Comment:
Do we ever expect this to occur? This doesn't seem foolproof because it's not looking at `DisconnectedCircuits`. Is this worth a comment that it's just here to prevent errors in unit tests?